### ISSUE TITLE
Unusable in .NET Core

### ISSUE THREAD

**Issue by Ranjit-S** [robinrodricks/FluentFTP#160]
I am unable to get listing nor able to upload a file to a remote FTP server in a ASP.NET Core project. Tested 
the same piece of code in a .NET 4.6 project and it works. FileZilla also has no issues uploading files.

In the .NET Core project, a file is created on the remote server but no data is transferred.
![capture](https://user-images.githubusercontent.com/8552516/29115013-62c2c2b4-7d39-11e7-9cdf-1c4989d0509a.PNG)

**Tested on the following workstations**
PC 1 and PC 2
OS: Windows 7 and Windows 10, respectively
FileZilla: 3.27.0.1
Firewall: Off
Visual Studio 2017
.NET Core: 1.1
FluentFTP: 17.6.1, also tested two previous versions

**Server**
Ubuntu: 16.04 LTS

**.NET Core**

    FtpTrace.LogToFile = "log.txt";
    FtpTrace.LogIP = false;
    FtpTrace.LogUserName = false;
    FtpTrace.LogPassword = false;

    FtpClient client = new FtpClient(HOST, USERNAME, PASSWORD);
    client.DataConnectionType = FtpDataConnectionType.AutoActive;
    client.DataConnectionEncryption = false;
    client.EncryptionMode = FtpEncryptionMode.None;
    client.Encoding = Encoding.UTF8;

    try
    {
        client.Connect();

        var list = client.GetListing("/");

        var data = Encoding.UTF8.GetBytes("Some data");
        client.Upload(data, "/test1.csv");
    }
    catch (Exception ex)
    {
    }

    if (client.IsConnected)
        client.Disconnect();

**Trace**
```

# Connect()
Status:   Connecting to ***:21
Response: 220 Please visit https://filezilla-project.org/
Response: 220-FileZilla Server 0.9.60 beta
Response: 220-written by Tim Kosse (tim.kosse@filezilla-project.org)
Command:  USER ***
Response: 331 Password required for abcd
Command:  PASS ***
Response: 230 Logged on
Command:  FEAT
Response: 211 End
Response: 211-Features:
Response: MDTM
Response: REST STREAM
Response: SIZE
Response: MLST type*;size*;modify*;
Response: MLSD
Response: UTF8
Response: CLNT
Response: MFMT
Response: EPSV
Response: EPRT
Status:   Text encoding: System.Text.UTF8Encoding
Command:  OPTS UTF8 ON
Response: 202 UTF8 mode is always enabled. No need to send this command.
Command:  SYST
Response: 215 UNIX emulated by FileZilla
Status:   Auto-detected UNIX listing parser

# GetListing("/", Auto)
Command:  TYPE I
Response: 200 Type set to I

# OpenActiveDataStream(AutoActive, "MLSD /", 0)
Command:  EPRT |1|10.0.0.32|50764|
Response: 200 Port command successful
Command:  MLSD /
Response: 150 Opening data channel for directory listing of "/"
+---------------------------------------+
-----------------------------------------
Status:   Disposing FtpSocketStream...

# Upload("/test1.csv", Overwrite, False)

# FileExists("/test1.csv")
Status:   There is stale data on the socket, maybe our connection timed out or you did not call GetReply(). Re-connecting...
Status:   The stale data was: 226 Successfully transferred "/"

# Connect()
Status:   Connecting to ***:21
Response: 220 Please visit https://filezilla-project.org/
Response: 220-FileZilla Server 0.9.60 beta
Response: 220-written by Tim Kosse (tim.kosse@filezilla-project.org)
Command:  USER ***
Response: 331 Password required for abcd
Command:  PASS ***
Response: 230 Logged on
Command:  FEAT
Response: 211 End
Response: 211-Features:
Response: MDTM
Response: REST STREAM
Response: SIZE
Response: MLST type*;size*;modify*;
Response: MLSD
Response: UTF8
Response: CLNT
Response: MFMT
Response: EPSV
Response: EPRT
Status:   Text encoding: System.Text.UTF8Encoding
Command:  OPTS UTF8 ON
Response: 202 UTF8 mode is always enabled. No need to send this command.
Command:  SYST
Response: 215 UNIX emulated by FileZilla
Status:   Auto-detected UNIX listing parser
Command:  SIZE /test1.csv
Response: 213 0

# DeleteFile("/test1.csv")
Command:  DELE /test1.csv
Response: 250 File deleted successfully

# OpenWrite("/test1.csv", Binary)
Command:  TYPE I
Response: 200 Type set to I

# GetFileSize("/test1.csv")
Command:  SIZE /test1.csv
Response: 550 File not found

# OpenActiveDataStream(AutoActive, "STOR /test1.csv", 0)
Command:  EPRT |1|10.0.0.32|50766|
Response: 200 Port command successful
Command:  STOR /test1.csv
Response: 150 Opening data channel for file upload to server of "/test1.csv"
Command:  QUIT
Response: 226 Successfully transferred "/test1.csv"

```

**.NET (4.6)**

    FtpTrace.LogIP = false;
    FtpTrace.LogUserName = false;
    FtpTrace.LogPassword = false;

    FtpClient client = new FtpClient(HOST, USERNAME, PASSWORD);
    client.DataConnectionType = FtpDataConnectionType.AutoActive;
    client.DataConnectionEncryption = false;
    client.EncryptionMode = FtpEncryptionMode.None;
    client.Encoding = Encoding.UTF8;

    try
    {
        client.Connect();

        var list = client.GetListing("/");

        var data = Encoding.UTF8.GetBytes("Some data");
        client.Upload(data, "/test.csv");

    }
    catch (Exception ex)
    {
    }

    if (client.IsConnected)
        client.Disconnect();

**Trace**

```# Connect()
Status:   Connecting to ***:21
Response: 220 Please visit https://filezilla-project.org/
Response: 220-FileZilla Server 0.9.60 beta
Response: 220-written by Tim Kosse (tim.kosse@filezilla-project.org)
Command:  USER ***
Response: 331 Password required for abcd
Command:  PASS ***
Response: 230 Logged on
Command:  FEAT
Response: 211 End
Response: 211-Features:
Response: MDTM
Response: REST STREAM
Response: SIZE
Response: MLST type*;size*;modify*;
Response: MLSD
Response: UTF8
Response: CLNT
Response: MFMT
Response: EPSV
Response: EPRT
Status:   Text encoding: System.Text.UTF8Encoding
Command:  OPTS UTF8 ON
Response: 202 UTF8 mode is always enabled. No need to send this command.
Command:  SYST
Response: 215 UNIX emulated by FileZilla
Status:   Auto-detected UNIX listing parser

# GetListing("/", Auto)
Command:  TYPE I
Response: 200 Type set to I

# OpenActiveDataStream(AutoActive, "MLSD /", 0)
Command:  EPRT |1|10.0.0.32|60090|
Response: 200 Port command successful
Command:  MLSD /
Response: 150 Opening data channel for directory listing of "/"
+---------------------------------------+
Listing:  type=dir;modify=20170809080806; Done
Listing:  type=dir;modify=20170809071909; Error
Listing:  type=dir;modify=20170809080801; Incoming
Listing:  type=dir;modify=20170809080806; Processing
Listing:  type=dir;modify=20170809075628; Response
-----------------------------------------

# CloseDataStream()
Response: 226 Successfully transferred "/"
Status:   Disposing FtpSocketStream...

# Upload("/test.csv", Overwrite, False)

# FileExists("/test.csv")
Command:  SIZE /test.csv
Response: 550 File not found

# OpenWrite("/test.csv", Binary)
Command:  TYPE I
Response: 200 Type set to I

# GetFileSize("/test.csv")
Command:  SIZE /test.csv
Response: 550 File not found

# OpenActiveDataStream(AutoActive, "STOR /test.csv", 0)
Command:  EPRT |1|10.0.0.32|60091|
Response: 200 Port command successful
Command:  STOR /test.csv
Response: 150 Opening data channel for file upload to server of "/test.csv"
Response: 226 Successfully transferred "/test.csv"
Command:  QUIT
Response: 221 Goodbye
Status:   Disposing FtpSocketStream...
Status:   Disposing FtpSocketStream...

# Dispose()
Status:   Disposing FtpClient object...
Status:   Disposing FtpSocketStream...
```

**Comment by Ranjit-S** [comment]
.NET Core trace was incomplete, apparently connect is called twice.

**Comment by hgupta9** [comment]
I don't know. More debugging is needed. .NET core is buggy and difficult to support in general.

**Comment by Ranjit-S** [comment]
I was going to debug the library in both frameworks, side by side. But was unable to compile the .NET project in VS 2017.

I'll try again later when I have time.

**Comment by hgupta9** [comment]
You need to use the VS2017 projects (`FluentFTP_Core14_VS2017.sln` and `FluentFTP_Core16_VS2017.sln`)

**Comment by artiomchi** [comment]
I think I just found the issue.. I've stumbled upon it myself, independently, while trying to get the unit tests to work, and realised that one of the tests was working on .NET 4.5, but was crashing when running it in .NET Standard.

Looking through the code, the issue is in FtpSocketStream / FtpDataStream. I'm not sure who wrote the code (if I understand it correctly, it's not @robinrodricks ), so I've got nobody to ask questions about why they did some things there :)

I'll be looking deeper into this, and will hopefully be able to provide a patch soon.
